Cookbook/Proposals/MealyReactiveFeedback.md

```diff
@@ -0,0 +1,141 @@
+
+## Motivation
+
+### 1. Allow `Feedback` to also handle "incoming event"
+
+Current `ReactiveFeedback` (as of ver 0.5.0) has `Feedback` type which represents side-effect and event feedback loop for a particular or entire system:
+
+```swift
+public struct Feedback<State, Event> {
+    let events: (Scheduler, Signal<State, NoError>) -> Signal<Event, NoError>
+}
+```
+
+(For simplicity, we will ignore `Scheduler` for this entire discussion)
+
+This kind of structure is so-called [Moore machine](https://en.wikipedia.org/wiki/Moore_machine) where feedback (output function) only takes `State` (or stream of `State`s) as argument and produces side-effects with accompanying new `Event` (or stream of `Event`s).
+
+Compared to this approach, current frontend architecture normally uses [Mealy machine](https://en.wikipedia.org/wiki/Mealy_machine), or [Queue automaton](https://en.wikipedia.org/wiki/Queue_automaton) (feedback model) instead, where reducer (transition function) takes both `State` and `Event`, then produces a new `State` and `Output` (with feedback).
+
+<details>
+<summary>Examples of Mealy machine (Click to open)</summary>
+
+For example, [Elm](https://package.elm-lang.org/packages/elm-lang/core/latest/Platform) uses `update: Event -> State -> (State, Cmd<Event>)` as reducer.
+Note that this single function can be splitted into two functions:
+
+1. `Event -> State -> State`, and
+2. `Event -> State -> Cmd<Event>`
+
+which we often call 1. reducer and 2. feedback.
+
+[Redux](https://redux.js.org) also provides similar mechanism called `Middleware`, where `middleware: (store: Store) -> (next: Dispatch) -> Action -> Any`.
+This can be rewritten as `State -> Observer<Event> -> Event -> Any` which can then be re-interpreted as `State -> Event -> Signal<Event>` (Caveat: This is not 100% logically correct).
+
+</details>
+
+More information about differences between Moore and Mealy can be found here:
+[Moore and Mealy Machines](https://www.tutorialspoint.com/automata_theory/moore_and_mealy_machines.htm)
+
+Please note that **both Moore and Mealy machines are theoretically convertible to each other** , so it is still reasonable to keep using current Moore-based machine.
+
+However, from practical point of view, Mealy machine has several advantages over Moore machine:
+
+- Requires less states
+    - This is a general conclusion of Mealy machine compared to Moore because we will add more graph-edges (transition patterns) than graph-nodes (states).
+- `Event`-recognizable `Feedback`
+    - For simple side-effects such as **logging**, **sending analytics events**, or somtimes **routing** or **image-loading** can be handled just from incoming events without precisely handling their states inside `ReactiveFeedback`
+    - This will remove verbose "idle -> routing -> idle" navigation state transition dance, and improves the feedback-based-navigation idea in [Babylonpartners/ios-playbook#44](https://github.com/Babylonpartners/ios-playbook/pull/44)
```

Comment:
>This will remove verbose "idle -> routing -> idle" navigation state transition dance

❤️ ❤️ ❤️

Reply:
IMO this is not just about getting rid of these intermediate states, it's about conceptually different way of reasoning about the state. IMO with this we will give up having the whole picture of state.

Comparing with approach that we used before, where routes (navigation signal) were derived from actions (events) rather than state and was kind of out of the feedback system, this change will effectively make it feel like it is _in_ the feedback system because it is installed as a feedback in a feedback loop rather than uses a separate subscription. But one could say that even with a separate subscription routes were still _in_ the feedback system because events were only produced by the system itself. So effectively to me it feels like it is going back to the previous way of defining routes but using a different syntax.

Reply:
> it feels like it is going back to the previous way of defining routes but using a different syntax.

This is right, though I'm not sure about the history of previous approach (how it didn't work well) since I wasn't there at that time.

My understanding is that #44 will introduce FRP's `flatMap(.first)` to also manage navigation state _indirectly_.
Since its FRP-based state can filter simultaneous multiple navigations, having extra navigating-states will mostly be redundant.

Reply:
> how it didn't work well

It worked fine, the problem was more theoretical than technical wether it's correct to derive routes from actions or from state.